### DISCUSSION TITLE
Does renovate support Gitlab CI variables in the renovate.json config file?

### DISCUSSION THREAD

**Discussion by ghost** [renovatebot/renovate#19307] [Request Help] [answered]
Hi guys,

I'm struggling a bit about something. I am hosting my own Gitlab server and Renovate. I use the package registry in Gitlab that requires an header Private-Token (or Deploy-Token or Job-Token) to authenticate and read from the Gitlab registry. 

I have managed to setup the authentication through the renovate.json file to send an Bearer token and read from the registry. So that works, yeah. Now the only problem is that the token in the hostRules section is plain text and I really want it to be an Gitlab CI variable, so that no one can see the personal access token of the renovate bot user.

I was wondering if someone has an solution for this. I put down an example:

**Renovate.json** works with token as plain text.
`{
   ...
  "hostRules": [
    {
      "matchHost": "<host>",
      "token": "token12345"
    }
 ...
}`

But I want to run it with an Gitlab CI variable, like this.
`{
  ...
  "hostRules": [
    {
      "matchHost": "<host>",
      "token": "${DEPLOYMENT_TOKEN}"
    }
  ],
  ...
}`

Hope someone can helps!

**Comment by ghost** [comment, marked as answer]
I already fixed it by putting the hostrules as an generic variable in the renovate environment variables like:
`RENOVATE_HOST_RULES: '[{"matchHost": "<host>","token": "${DEPLOYMENT_TOKEN}"}]'`

**Reply by JMLX42** [reply]
`token` did not work for me, but `password` worked.

Here is the CI variable I am using:

- name: `RENOVATE_HOST_RULES`
- value: `[{"matchHost": "${CI_REGISTRY}","username": "${GITLAB_USER_NAME}","password": "${RENOVATE_TOKEN}"}]`

Since it leverages only [GitLab CI predefined variables](https://docs.gitlab.com/ee/ci/variables/predefined_variables.html) (and `RENOVATE_TOKEN` documented in [renovate-runner](https://gitlab.com/renovate-bot/renovate-runner/)), it should work as-is for anyone with a self-hosted GitLab.

**Comment by viceice** [comment]
renovate will never replace any ENV vars inside JSON files. you should use a config.js file and normal JavaScript. there are plenty of samples here inside discussions.
